# Notebook 9 – Grouping & Aggregation



#1. groupby()

## Concept

`groupby()` is used to divide data into groups based on one or more columns.

After creating groups, we can perform calculations such as mean, sum, count, minimum, or maximum on each group.

## Simple Example

Suppose we have student data from different cities.

We can group the students according to their city and calculate the average marks of students from each city.

## Business Example

A company can group sales according to different regions and calculate total sales for each region.

## AI/ML Use Case

`groupby()` is useful during Exploratory Data Analysis (EDA) to understand patterns and statistics for different categories.

In [1]:
import pandas as pd

df = pd.DataFrame({
    "Name": ["Sourav", "Rahul", "Amit", "Riya", "Karan", "Priya"],
    "City": ["Kolkata", "Delhi", "Kolkata", "Delhi", "Mumbai", "Mumbai"],
    "Marks": [85, 90, 75, 88, 70, 95],
    "Gender": ["M", "M", "M", "F", "M", "F"]
})

df

,Name,City,Marks,Gender
0,Sourav,Kolkata,85,M
1,Rahul,Delhi,90,M
2,Amit,Kolkata,75,M
3,Riya,Delhi,88,F
4,Karan,Mumbai,70,M
5,Priya,Mumbai,95,F


In [2]:
df.groupby("City")["Marks"].mean()

,Marks
City,
Delhi,89.0
Kolkata,80.0
Mumbai,82.5


## Explanation

`groupby("City")` creates separate groups for each city.

Then:

`["Marks"].mean()`

calculates the average Marks for each city.

So instead of calculating one average for the complete dataset, we get a separate average for every city.

#2. agg()

## Concept

`agg()` is used when we want to perform multiple aggregation operations on grouped data.

For example, instead of calculating only the average, we can calculate the minimum, maximum, and average together.

## Simple Example

We can group students by City and calculate:

- Average marks
- Minimum marks
- Maximum marks

## Business Example

A company can group sales by region and calculate total sales, average sales, minimum sales, and maximum sales.

## AI/ML Use Case

`agg()` is useful in EDA and feature engineering when we need multiple statistical summaries from grouped data.

In [3]:
df.groupby("City")["Marks"].agg(["mean", "min", "max"])

,mean,min,max
City,,,
Delhi,89.0,88,90
Kolkata,80.0,75,85
Mumbai,82.5,70,95


## Explanation

`agg()` allows us to apply multiple functions at the same time.

Here:

- `mean` → average marks
- `min` → lowest marks
- `max` → highest marks

This gives us a detailed summary for every city.

#3. transform()

## Concept

`transform()` performs a calculation within each group but returns the result with the same number of rows as the original DataFrame.

This is an important difference from `groupby().mean()`, which reduces the number of rows.

## Simple Example

Suppose we want to calculate the average marks of each city and add that average beside every student belonging to that city.

## Business Example

A company can calculate the average salary of each department and show that department average beside every employee.

## AI/ML Use Case

`transform()` can be useful for feature engineering because the resulting values can be added directly as a new feature to the original dataset.

In [4]:
df["City_Average"] = df.groupby("City")["Marks"].transform("mean")

df

,Name,City,Marks,Gender,City_Average
0,Sourav,Kolkata,85,M,80.0
1,Rahul,Delhi,90,M,89.0
2,Amit,Kolkata,75,M,80.0
3,Riya,Delhi,88,F,89.0
4,Karan,Mumbai,70,M,82.5
5,Priya,Mumbai,95,F,82.5


## Explanation

`groupby("City")` creates groups based on City.

`transform("mean")` calculates the average Marks for each city.

The important part is that the result has the same number of rows as the original DataFrame.

For example, both Kolkata students receive the Kolkata average of 80.

#4. filter()

## Concept

`filter()` is used with grouped data to keep or remove complete groups based on a condition.

Unlike normal row filtering, here we can decide whether an entire group should remain in the result.

## Simple Example

Suppose we want to keep only cities where the average marks are greater than 80.

## Business Example

A company may want to keep only sales regions whose average sales are above a certain target.

## AI/ML Use Case

Grouped filtering can be useful during data preprocessing when we want to remove categories or groups that do not have enough useful data.

In [5]:
result = df.groupby("City").filter(
    lambda group: group["Marks"].mean() > 80
)

result

,Name,City,Marks,Gender,City_Average
1,Rahul,Delhi,90,M,89.0
3,Riya,Delhi,88,F,89.0
4,Karan,Mumbai,70,M,82.5
5,Priya,Mumbai,95,F,82.5


## Explanation

The function calculates the average Marks for each city.

If the average is greater than 80, the complete group is kept.

In our dataset:

- Kolkata → 80 → removed because it is not greater than 80
- Delhi → 89 → kept
- Mumbai → 82.5 → kept

Therefore, all students from Delhi and Mumbai remain.

#5. Pivot Tables

## Concept

A Pivot Table is used to summarize and organize data based on different categories.

It allows us to compare values across groups in a compact table.

## Simple Example

We can create a Pivot Table showing the average marks for each city and gender.

## Business Example

A company can create a Pivot Table to compare sales across different regions and product categories.

## AI/ML Use Case

Pivot Tables are useful during EDA because they help us quickly understand relationships between categorical variables and numerical values.

In [6]:
pivot = pd.pivot_table(
    df,
    values="Marks",
    index="City",
    columns="Gender",
    aggfunc="mean"
)

pivot

Gender,F,M
City,,
Delhi,88.0,90.0
Kolkata,NaN,80.0
Mumbai,95.0,70.0


## Explanation

Here:

- `values="Marks"` → the numerical value we want to summarize
- `index="City"` → rows are grouped by City
- `columns="Gender"` → columns are divided by Gender
- `aggfunc="mean"` → average Marks are calculated

This creates a summary table that makes comparison easier.

#6. Crosstab

## Concept

`crosstab()` is used to create a frequency table showing how often combinations of categorical values occur.

It is mainly useful for counting relationships between categorical variables.

## Simple Example

We can use `crosstab()` to count how many male and female students belong to each city.

## Business Example

A company can use a crosstab to count customers according to gender and region.

## AI/ML Use Case

Crosstabs are useful during EDA for understanding relationships between categorical features.

In [7]:
pd.crosstab(df["City"], df["Gender"])

Gender,F,M
City,,
Delhi,1,1
Kolkata,0,2
Mumbai,1,1


## Explanation

`crosstab()` counts the combinations of two categorical columns.

Here:

- Rows represent City.
- Columns represent Gender.
- Values represent the number of students in each combination.

This helps us quickly understand the distribution of categorical data.